### PR TITLE
Fixes #11484

### PR DESCRIPTION
**Pull request by codingdave**
- Replaced redundant function (_string_get_next_variable) with simplified _string_get_next_variable
- Completely rewritten _string_get_next_variable fixes the former segmentation fault

Howto formerly trigger the segmentation fault:
1) Have a SD card in the card reader with DCIM structure and images
2) Open dt
3) Click on Import > Import from camera
4) In the UI "Import images from camera" select any image and click on import -> Segmentation fault
SUMMARY: AddressSanitizer: heap-buffer-overflow src/common/variables.c:79 in _string_get_next_variable

### CODE REVIEW

File: src/common/variables.c [hunk 1]
```diff
@@ -39,59 +39,23 @@ typedef struct dt_variables_data_t
...
-        strncpy(variable, p, e - p + 1);
-        variable[e - p + 1] = '\0';
-        return p + 1;
-      }
-      else
-        return NULL;
-    }
-  }
-  return NULL;
+	gchar *end = NULL;
```
Comment (LebedevRI) [inline src/common/variables.c]: Oh, are those tabs?
Which editor are you using, that the modelines didn't take any effect?

File: src/common/variables.c [hunk 2]
```diff
@@ -39,59 +39,23 @@ typedef struct dt_variables_data_t
...
+	gchar *start = g_strstr_len(string, strlen(string), "$(");
+	if(start)
+	{
+		end = g_strstr_len(start, strlen(start), ")");
+		if(end)
+		{
+			const int length = end - start + 1;
+			strncpy(variable, start, length);
```
Comment (LebedevRI) [inline src/common/variables.c]: Please use `g_strlcpy()`, it makes sure that there is no overflow, and the string is zero-terminated in all cases.

Comment (LebedevRI) [inline src/common/variables.c]: I'd also try to avoid `if() if()` by inverting the check and returning early.

File: src/common/variables.c [hunk 3]
```diff
@@ -351,42 +315,48 @@ gchar *dt_variables_get_result(dt_variables_params_t *params)
 
 gboolean dt_variables_expand(dt_variables_params_t *params, gchar *string, gboolean iterate)
 {
-  gchar *variable = g_malloc(128);
-  gchar *value = g_malloc_n(1024, sizeof(gchar));
-  gchar *token = NULL;
+	gchar *variable = g_malloc(128);
+	gchar *value = g_malloc_n(1024, sizeof(gchar));
```
Comment (LebedevRI) [inline src/common/variables.c]: Since `  ` (2 spaces) were replaced with `	` (tab), i'm not sure what is actually changed here...

**Comment by codingdave** [comment]
- Overflow situation improved
- more easy to read by coding with clear intention
- comments where needed
- The code is working again

In contrast to https://gist.github.com/LebedevRI/99d6d9753f2f02e5937fbea663307a45 the code has the idea to walk along the complete string with 2 pointers: `char* string` is always pointing to the next subtoken in the input string as input for the function. From here `"(` and `)` are being looked for. The position after `)` needs to be the return value of the function because it will be `string` input for the next iterative function call. The range including `$(` and `)` is copied to the resulting `variable` and is used for variable expansion.  

@LebedevRI I think the function was not easy to grasp and this is a useful comment. Whether I add this, however, is up to you.

File: src/common/variables.c [hunk 4]
```diff
@@ -39,59 +39,26 @@ typedef struct dt_variables_data_t
   guint sequence;
 } dt_variables_data_t;
 
-gchar *_string_get_first_variable(gchar *string, gchar *variable)
+gchar *_string_get_next_variable(gchar *string, gchar *variable, const size_t variable_size)
 {
-  if(g_strrstr(string, "$("))
+  // initiate empty output variable
+  *variable = '\0';
```
Comment (LebedevRI) [inline src/common/variables.c]: this is meaningless assignment

**Pull request merged**